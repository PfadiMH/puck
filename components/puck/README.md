# /components/puck

This directory contains React components specifically designed for use as building blocks within the Puck editor.

**Key points:**

- **Important:** All components must be registered in `lib/config/page.config.ts` to be available in the Puck editor.
- [Puck Docs on Defining Components](https://puckeditor.com/docs/integrating-puck/component-configuration)

### Component Configuration

Each component within this directory must export two key items for it to be registered and utilized in `page.config.ts`: a `config` object and a `config` type.

This `config` specifies how the component should be rendered and what fields are available for editing.

- `render`: This property is a function that returns the React component to be rendered. It receives the props as defined in the `fields` object.
- `fields`: This is an object that defines the editable fields for the component in the Puck editor. The value is an object that specifies the type of input to be used in the editor (e.g., "text", "select").
