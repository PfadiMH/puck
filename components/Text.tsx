import { ComponentConfig } from "@measured/puck";
import React from "react";

export type TextProps = {
  text: string;
};

function Text({ text }: TextProps) {
  return <p className="font-sans text-wrap py-3">{text}</p>;
}

export const textConfig: ComponentConfig<TextProps> = {
  render: Text,
  fields: {
    text: {
      type: "textarea",
    },
  },
  defaultProps: {
    text: "Mir sind Voll DÃ¤bii!",
  },
};
