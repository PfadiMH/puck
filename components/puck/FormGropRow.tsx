import { Input } from "@components/misc/external/input";
import { FormField } from "./FormGroup";

export function FormGroupRow({ row }: { row: FormField[] }) {
  return (
    <div className='flex'>
      {row.map((field, index) => (
        <span className='flex grow mt-2 mb-2' key={`field-${index}`}>
          <span className='flex grow'>
            <span
              className={`${field.size === "full" ? "flex-1" : "flex-1/2"} ${
                index === 0 ? "mr-2" : ""
              }`}
            >
              <label htmlFor={field.label}>{field.label}</label>
              <Input
                id={field.label}
                name={field.label}
                placeholder={field.placeholder}
                required={field.required}
                type={field.type}
              />
            </span>
          </span>
        </span>
      ))}
    </div>
  );
}
