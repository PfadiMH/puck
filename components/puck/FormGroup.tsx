import { Input } from "@components/misc/external/input";
import { Label } from "@components/misc/external/label";
import FormGroupForm from "@components/page/FormGroupForm";
import { ComponentConfig, WithId } from "@measured/puck";

export type FormGroupProps = {
  formFields: { label: string; size: string }[];
  buttonLabel: string;
  formRecipientEmail: string;
};

export function FormGroup({
  formFields,
  buttonLabel,
  id,
}: WithId<FormGroupProps>) {
  const groupedFields: { label: string; size: string }[][] = [];
  let tempRow: { label: string; size: string }[] = [];
  formFields.forEach((field, index) => {
    if (field.size === "full") {
      groupedFields.push([field]);
    } else if (field.size === "half") {
      if (tempRow.length === 0) {
        tempRow.push(field);
        if (index === formFields.length - 1) {
          groupedFields.push(tempRow);
        }
      } else if (tempRow.length === 1) {
        tempRow.push(field);
        groupedFields.push(tempRow);
        tempRow = [];
      }
    }
  });
  return (
    <FormGroupForm componentId={id}>
      {groupedFields.map((row) => (
        <div className='flex'>
          {row.map((field, index) => (
            <span className='flex grow mt-2 mb-2'>
              {field.size === "full" ? (
                <span className='flex grow flex-col'>
                  <Label htmlFor={field.label}>{field.label}</Label>
                  <Input
                    id={field.label}
                    name={field.label}
                    placeholder={field.label}
                  />
                </span>
              ) : (
                <span className='flex grow'>
                  <span className='flex-1/2'>
                    <Label htmlFor={field.label}>{field.label}</Label>
                    <Input
                      id={field.label}
                      name={field.label}
                      placeholder={field.label}
                    />
                  </span>
                  {index === 0 ? (
                    <span>
                      {row.length - 1 !== index ? (
                        <span className='p-1'></span>
                      ) : (
                        <span></span>
                      )}
                    </span>
                  ) : (
                    <span></span>
                  )}
                </span>
              )}
            </span>
          ))}
        </div>
      ))}
      {groupedFields.length > 0 ? (
        <button
          className='bg-gradient-to-br relative group/btn from-black dark:from-zinc-900 dark:to-zinc-900 to-neutral-600 block dark:bg-zinc-800 w-full text-white rounded-md h-10 font-medium shadow-[0px_1px_0px_0px_#ffffff40_inset,0px_-1px_0px_0px_#ffffff40_inset] dark:shadow-[0px_1px_0px_0px_var(--zinc-800)_inset,0px_-1px_0px_0px_var(--zinc-800)_inset]'
          type='submit'
        >
          {buttonLabel}
        </button>
      ) : (
        <span></span>
      )}
    </FormGroupForm>
  );
}

export const formGroupConfig: ComponentConfig<FormGroupProps> = {
  render: FormGroup,
  fields: {
    formFields: {
      type: "array",
      arrayFields: {
        label: {
          type: "text",
        },
        size: {
          type: "radio",
          options: [
            { value: "full", label: "full" },
            { value: "half", label: "half" },
          ],
        },
      },
      defaultItemProps: {
        label: "label",
        size: "full",
      },
    },
    buttonLabel: {
      type: "text",
    },
    formRecipientEmail: {
      type: "text",
    },
  },
  defaultProps: {
    formFields: [{ label: "label", size: "full" }],
    buttonLabel: "Send response",
    formRecipientEmail: "example@example.com",
  },
};
