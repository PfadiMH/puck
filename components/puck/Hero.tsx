import { PostHeroSvg } from "@components/graphics/PostHeroSvg";
import {
  FileProps,
  singleSelectionFileTableField,
} from "@components/puck-fields/fileTable";
import { ComponentConfig } from "@measured/puck";
import Image from "next/image";

export type HeroProps = {
  title: string;
  backgroundImage: FileProps;
};

function Hero({ title, backgroundImage }: HeroProps) {
  const imageUrl = backgroundImage?.url;

  return (
    <div className="full w-full h-96 relative flex flex-col justify-center overflow-hidden items-center">
      {imageUrl && (
        <Image
          fill
          src={imageUrl}
          alt="Hero Image"
          style={{ objectFit: "cover" }}
        />
      )}
      {title && (
        <>
          <h1 className="text-4xl font-bold text-center z-10 text-white">
            {title}
          </h1>
          <div className="bg-black opacity-15 absolute w-full h-full" />
        </>
      )}
      <PostHeroSvg className="absolute bottom-0" />
    </div>
  );
}

export const heroConfig: ComponentConfig<HeroProps> = {
  render: Hero,
  fields: {
    title: {
      type: "text",
      label: "Title (Optional)",
    },
    backgroundImage: singleSelectionFileTableField,
  },
};
