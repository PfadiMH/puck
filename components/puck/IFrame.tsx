import { ComponentConfig } from "@measured/puck";
import React from "react";

export type IFrameProps = {
  source: string;
  height: string;
  title: string;
};

function IFrame({ source, height, title }: IFrameProps) {
  return (
    <>
      <div className="w-full overflow-hidden">
        <iframe src={source} width="100%" height={height} title={title} />
      </div>
    </>
  );
}

export const iframeConfig: ComponentConfig<IFrameProps> = {
  render: IFrame,
  fields: {
    source: {
      type: "textarea",
    },
    height: {
      type: "number",
    },
    title: {
      type: "textarea",
    },
  },
};
