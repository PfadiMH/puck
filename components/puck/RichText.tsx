import { RichTextEditor } from "@components/misc/RichTextEditor";
import {
  defaultRichTextValue,
  richTextField,
} from "@components/puck-fields/rich-text";
import { ComponentConfig, WithPuckProps } from "@measured/puck";
import { JSONContent } from "@tiptap/react";

export type RichTextProps = {
  content: JSONContent;
};

function RichText({ content }: WithPuckProps<RichTextProps>) {
  return <RichTextEditor readonly content={content} />;
}

export const richTextConfig: ComponentConfig<RichTextProps> = {
  render: RichText,
  label: "Rich Text",
  fields: {
    content: richTextField,
  },
  defaultProps: {
    content: defaultRichTextValue,
  },
};
