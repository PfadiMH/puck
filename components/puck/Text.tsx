import { ComponentConfig } from "@measured/puck";

export type TextProps = {
  text: string;
};

function Text({ text }: TextProps) {
  return <p className="text-wrap">{text}</p>;
}

export const textConfig: ComponentConfig<TextProps> = {
  render: Text,
  fields: {
    text: {
      type: "textarea",
    },
  },
  defaultProps: {
    text: "Mir sind Voll DÃ¤bii!",
  },
};
