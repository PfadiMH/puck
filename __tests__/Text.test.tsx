import { textConfig } from "@components/puck/Text";
import { PuckContext } from "@measured/puck";
import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";

describe("Text component", () => {
  it("renders the component unchanged", () => {
    const { container } = render(
      <textConfig.render
        {...{
          id: "test-id",
          ...textConfig.defaultProps!,
          puck: {} as PuckContext,
        }}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it("renders the component with default props from the config", () => {
    render(
      <textConfig.render
        {...{
          id: "test-id",
          ...textConfig.defaultProps!,
          puck: {} as PuckContext,
        }}
      />
    );

    const textElement = screen.getByText(textConfig.defaultProps!.text);
    expect(textElement).toBeInTheDocument();
  });

  it("renders the component with custom props", () => {
    const customText = "This is a custom text";
    render(
      <textConfig.render
        {...{
          id: "test-id",
          text: customText,
          puck: {} as PuckContext,
        }}
      />
    );

    const textElement = screen.getByText(customText);
    expect(textElement).toBeInTheDocument();
  });
});
