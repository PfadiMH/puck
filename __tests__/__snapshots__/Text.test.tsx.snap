// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Text component renders the component unchanged 1`] = `
<div>
  <p
    class="text-wrap"
  >
    Mir sind Voll DÃ¤bii!
  </p>
</div>
`;
